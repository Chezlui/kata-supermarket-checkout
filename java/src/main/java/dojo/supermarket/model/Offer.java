package dojo.supermarket.model;

public interface Offer {
    Product getProduct();

    Discount getDiscount(Product product, double quantity, double priceForOne);
}
