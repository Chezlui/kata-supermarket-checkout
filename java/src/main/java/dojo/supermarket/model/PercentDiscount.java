package dojo.supermarket.model;

public class PercentDiscount extends Offer {
    private final Product product;
    private final double percentDiscount;

    public PercentDiscount(Product product, double percentDiscount) {
        this.product = product;
        this.percentDiscount = percentDiscount;
    }

    @Override
    public Product getProduct() {
        return product;
    }

    @Override
    public Discount getDiscount(Product product, double quantity, double priceForOne) {
        return new Discount( product, this.percentDiscount + "% off", quantity*priceForOne*this.percentDiscount/100.0);
    }
}
