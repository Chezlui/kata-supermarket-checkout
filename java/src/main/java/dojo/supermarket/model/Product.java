package dojo.supermarket.model;

public class Product {
}
