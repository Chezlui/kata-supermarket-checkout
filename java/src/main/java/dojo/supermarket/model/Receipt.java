package dojo.supermarket.model;

public class Receipt {
    public Double getTotalPrice() {
        return null;
    }
}
