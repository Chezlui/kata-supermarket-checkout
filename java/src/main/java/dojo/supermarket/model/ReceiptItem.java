package dojo.supermarket.model;

import java.util.Objects;

public class ReceiptItem {
    private final Product product;
    private final double price;
    private final int quantity;

    public ReceiptItem(Product p, int quantity, double price) {
        this.product = p;
        this.quantity = quantity;
        this.price = price;
    }

    public double getPrice() {
        return this.price;
    }

    public Product getProduct() {
        return product;
    }

    public int getQuantity() {
        return quantity;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ReceiptItem that = (ReceiptItem) o;
        return Double.compare(that.price, price) == 0 &&
                quantity == that.quantity &&
                Objects.equals(product, that.product);
    }

    @Override
    public int hashCode() {

        return Objects.hash(product, price, quantity);
    }
}
