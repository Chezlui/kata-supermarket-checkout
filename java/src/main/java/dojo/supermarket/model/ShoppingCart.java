package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ShoppingCart {

    private final List<Product> items = new ArrayList<>();

    List<Product> getItems() {
        return new ArrayList<>(items);
    }

    void addItem(Product product) {
        items.add(product);
    }

    Map<Product, Integer> productQuantities() {
        Map<Product, Integer> productQuantities = new HashMap<>();
        for (Product p : getItems()) {
            if (productQuantities.containsKey(p)) {
                productQuantities.put(p, productQuantities.get(p) + 1);
            } else {
                productQuantities.put(p, 1);
            }
        }
        return productQuantities;
    }


}
