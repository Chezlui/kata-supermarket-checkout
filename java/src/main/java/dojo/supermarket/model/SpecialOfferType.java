package dojo.supermarket.model;

public enum SpecialOfferType {
    ThreeForTwo, PercentDiscount, XForY;
}
