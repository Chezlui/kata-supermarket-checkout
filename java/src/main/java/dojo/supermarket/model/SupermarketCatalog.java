package dojo.supermarket.model;

public interface SupermarketCatalog {
}
