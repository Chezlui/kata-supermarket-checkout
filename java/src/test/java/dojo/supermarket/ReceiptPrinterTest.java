package dojo.supermarket;

import dojo.supermarket.model.*;
import org.approvaltests.Approvals;
import org.junit.Test;

public class ReceiptPrinterTest {

    Product toothbrush = new Product("toothbrush", ProductUnit.Each);
    Product apples = new Product("apples", ProductUnit.Kilo);

    @Test
    public void oneLineItem() {
        ReceiptItem item = new ReceiptItem(toothbrush, 1, 0.99, 0.99);
        Approvals.verify(new ReceiptPrinter(40).printReceiptLineItem(item));
    }

    @Test
    public void quantityTwo() {
        ReceiptItem item = new ReceiptItem(toothbrush, 2, 0.99,0.99 * 2);
        Approvals.verify(new ReceiptPrinter(40).printReceiptLineItem(item));
    }

    @Test
    public void looseWeight() {
        ReceiptItem item = new ReceiptItem(apples, 2.3, 1.99,1.99 * 2.3);
        Approvals.verify(new ReceiptPrinter(40).printReceiptLineItem(item));
    }

    @Test
    public void total() {

        Receipt receipt = new Receipt();
        receipt.addProduct(toothbrush, 1, 0.99, 2*0.99);
        receipt.addProduct(apples, 0.75, 1.99, 1.99*0.75);
        Approvals.verify(new ReceiptPrinter(40).printTotal(receipt));
    }

    @Test
    public void discounts() {
        Receipt receipt = new Receipt();
        receipt.addDiscount(new Discount(apples, "3 for 2", 0.99));
        Approvals.verify(new ReceiptPrinter(40).printDiscounts(receipt));
    }

    @Test
    public void printWholeReceipt() {
        Receipt receipt = new Receipt();
        receipt.addProduct(toothbrush, 1, 0.99, 0.99);
        receipt.addProduct(toothbrush, 2, 0.99, 2*0.99);
        receipt.addProduct(apples, 0.75, 1.99, 1.99*0.75);
        receipt.addDiscount(new Discount(toothbrush, "3 for 2", 0.99));
        Approvals.verify(new ReceiptPrinter(40).printReceipt(receipt));
    }

}
