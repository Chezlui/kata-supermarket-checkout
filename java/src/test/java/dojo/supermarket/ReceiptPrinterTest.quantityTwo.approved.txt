toothbrush                          1.98
  0.99 * 2
